### PR TITLE
[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py, api.py and init_state_dict()

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* #88636
* #88635
* __->__ #88481
* #87900

**Motivation**:
Several Enums, Dataclasses and states defined in fully_sharded_data_paralle.py should be moved to a place where the composable FSDP can access. This PR does the move.

### CODE REVIEW

**Comment by awgu** [comment]
I am wondering if it is possible to / if we should move the public-facing state dict classes to `api.py`. I am pretty novice at figuring out these imports and public bindings though.

What would the import look like if a user wanted to get a `StateDictConfig` or `StateDictType`? In my understanding, we should not have users import from `torch.distributed.fsdp._state_dict_utils`. (That was why I moved classes like `BackwardPrefetch`, `ShardingStrategy`, etc. that the user interacts with to `api.py`. Classes like `TrainingState` which are only used internally are in `_common_utils.py`.)

**Comment by fegin** [comment]
@awgu Users should not use `torch.distributed.fsdp._state_dict_utils` but should use `torch.distributed.fsdp`. Any public interfaces should be import in `torch.distributed.fsdp.__init__.py`. Then it won't matter where are those interfaces are defined internally. But I'm open to move to `api.py` if you prefer.

**Comment by awgu** [comment]
> @awgu Users should not use `torch.distributed.fsdp._state_dict_utils` but should use `torch.distributed.fsdp`. Any public interfaces should be import in `torch.distributed.fsdp.__init__.py`. Then it won't matter where are those interfaces are defined internally. But I'm open to move to `api.py` if you prefer.

This makes sense. (No need to move to `api.py`.) I think everything should be fine since both `_composable/fully_shard.py` and `fsdp/fully_sharded_data_parallel.py` can import from `_state_dict_utils.py` without issue. Plus, we expose in `fsdp/__init__.py`.

**Review by rohan-varma — APPROVED** [review]
LG

**Review by awgu — APPROVED** [review]
LGTM!

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 1]
```diff
@@ -71,7 +100,10 @@ def _enter_full_param_ctx(
 
 def _exit_full_param_ctx(module) -> None:
     """A helper function to exit ``summon_full_params`` context."""
-    module._assert_state([TrainingState.SUMMON_FULL_PARAMS])
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/fsdp/_state_dict_utils.py:74]: Using `TrainingState` is problematic for state_dict_utils as composable FSDP only has one training state. Using handle state makes more sense and will be done in the following PRs.

**Pull request closed without merging**